File: src/main/java/dragonbook/lexer/Word.java
package dragonbook.lexer;

public interface Word extends Token {

    String lexeme();

    Word AND = new Keyword("&&", Tag.AND);
    Word OR = new Keyword("||", Tag.OR);
    Word EQ = new Keyword("==", Tag.EQ);
    Word NE = new Keyword("!=", Tag.NE);
    Word LE = new Keyword("<=", Tag.LE);
    Word GE = new Keyword(">=", Tag.GE);
    Word MINUS = new Keyword("minus", Tag.MINUS);
    Word TRUE = new Keyword("true", Tag.TRUE);
    Word FALSE = new Keyword("false", Tag.FALSE);
    Word TEMP = new Keyword("t", Tag.TEMP);
    Word IF = new Keyword("if", Tag.IF);
    Word ELSE = new Keyword("else", Tag.ELSE);
    Word WHILE = new Keyword("while", Tag.WHILE);
    Word DO = new Keyword("do", Tag.DO);
    Word BREAK = new Keyword("break", Tag.BREAK);
    Word ACCESS = new Keyword("[]", Tag.INDEX);
}

File: src/main/kotlin/dragonbook/lexer/Word.kt
package dragonbook.lexer

interface Word : Token {
    val lexeme: String

    companion object {
        val AND: Word = Keyword("&&", Tag.AND)
        val OR: Word = Keyword("||", Tag.OR)
        val EQ: Word = Keyword("==", Tag.EQ)
        val NE: Word = Keyword("!=", Tag.NE)
        val LE: Word = Keyword("<=", Tag.LE)
        val GE: Word = Keyword(">=", Tag.GE)
        val MINUS: Word = Keyword("minus", Tag.MINUS)
        val TRUE: Word = Keyword("true", Tag.TRUE)
        val FALSE: Word = Keyword("false", Tag.FALSE)
        val TEMP: Word = Keyword("t", Tag.TEMP)
        val IF: Word = Keyword("if", Tag.IF)
        val ELSE: Word = Keyword("else", Tag.ELSE)
        val WHILE: Word = Keyword("while", Tag.WHILE)
        val DO: Word = Keyword("do", Tag.DO)
        val BREAK: Word = Keyword("break", Tag.BREAK)
        val INDEX: Word = Keyword("[]", Tag.INDEX)
    }
}

data class Name(override val lexeme: String) : Word {
    override val tag: Int get() = Tag.ID
    override fun toString(): String = lexeme
}

data class Keyword(override val lexeme: String, override val tag: Int) : Word {
    override fun toString(): String = lexeme
}